### PR TITLE
PHPArray adapter using arrays for hierarchical

### PR DESCRIPTION
**Pull request by Korbeil**
| Question      | Answer
| ------------- | ------
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

### Description

Replacing plain array for PHPArray adapter and using brand new array architecture to store cached values. Mostly improve how we manage Hierarchical cache since we'll base onto nested arrays.

### TODO
* [x] Add tests
* [x] Updated Changelog.md

### CODE REVIEW

**Comment by Korbeil** [comment]
Possible to re-run PHP 5.6 low deps ? I think we got a problem with that curl download :)

File: src/Hierarchy/HierarchicalArrayCachePoolTrait.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Cache\Hierarchy;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+trait HierarchicalArrayCachePoolTrait
```
Comment (Nyholm) [inline src/Hierarchy/HierarchicalArrayCachePoolTrait.php]: This trait is only used once. Lets not have a trait and instead put in in the ArrachCachePool class.

Comment (Korbeil) [reply src/Hierarchy/HierarchicalArrayCachePoolTrait.php]: Was planning to use it for the nested directories for Filesystem Adapter ;)

Comment (Korbeil) [reply src/Hierarchy/HierarchicalArrayCachePoolTrait.php]: @Nyholm any updates ?

Comment (Nyholm) [reply src/Hierarchy/HierarchicalArrayCachePoolTrait.php]: It is just one function and this trait is using another trait. 

Im hesitant about this. Understanding the hierarchy cache is difficult enough. By removing this trait and possible duplicate the 3 lines of code it's providing will make things easier to read and understand

File: src/Hierarchy/Changelog.md [hunk 2]
```diff
@@ -4,6 +4,10 @@ The change log describes what is "Added", "Removed", "Changed" or "Fixed" betwee
 
 ## UNRELEASED
 
+### Added
+
+* New trait: `HierarchicalArrayCachePoolTrait` to manage keys as array for some custom implementations.
```
Comment (Nyholm) [inline src/Hierarchy/Changelog.md]: This is not needed when the trait is removed.

**Comment by Korbeil** [comment]
@Nyholm should be all good. About tests, there is already the `Cache\Adapter\Illuminate\Tests\IntegrationHierarchyTest` class handling Hierarchy ?

**Comment by Nyholm** [comment]
Yeah, you basically only need to "enable" tests. See https://github.com/php-cache/cache/blob/db6e5af7735c494052e40e8d1042969eff548e9e/src/Adapter/Redis/Tests/IntegrationHierarchyTest.php

**Comment by Nyholm** [comment]
Thank you. I'll just wait for traivs

**Pull request merged**